realtek: add support for the Netgear GS108T v3

Musashii-kun opened the pull request:
The GS108T is a cheap 8-port network-managed gigabit switch.  This change is essentially a refactor of the GS110TPP v1 support code.

The Netgear open-source GPL archive for these two switches is identical.  The FW magic for official firmware images matches as well.

svanheule commented on target/linux/realtek/dts/rtl8380_netgear_gs108t-v3.dts:
```diff
@@ -0,0 +1,202 @@
+			partition@300000{
+				label = "firmware";
+				compatible = "netgear,uimage";
+				reg = <0x0300000 0x1d00000>;
```
This merges the two firmware partitions for more space, but makes us loose the ability to dual-boot. Two partitions half the size is probably enough for most use-cases now. (You probably copied this from the gs110tpp, and I have locally reverted this partition merge myself)

Rupan replied:
The change is applied as suggested.

svanheule replied:
One of the partitions will still need to have ```label = "firmware"```. I don't think OpenWrt will recognize it as the firmware partition otherwise.

svanheule commented on target/linux/realtek/dts/rtl8380_netgear_gs108t-v3.dts:
```diff
@@ -0,0 +1,202 @@
+		#size-cells = <0>;
+
+		/* Internal phy */
+		phy8: ethernet-phy@8 {
```
rtl838x.dtsi has a few macros you can use to make this more concise.

Rupan replied:
Done, for this and the switch ports as well.

svanheule commented on target/linux/realtek/dts/rtl8380_netgear_gs108t-v3.dts:
```diff
@@ -0,0 +1,202 @@
+
+// Memory
+// ------
+// Spec: 256 MB DDR3-1600
```
Are you sure about this? I recently discovered my mistake for the GS110TPP, which only has 128MB or RAM.

As @bkobl noted, this info should (also) go into the commit message. Since some Realtek switches have one MAC-address per port, and others (like this one) have one MAC-address for the entire device, this should probably also be mentioned in the commit message.

Rupan replied:
Commented below, but adding more context here.  The datasheet for the chip indicates that it contains 256MB of memory.  However, U-Boot reports 128MB of memory at boot time.  I have a suspicion that something in the SoC artificially limits the exposed memory to upsell "pro" versions of the same board.  This may require reverse engineering the U-Boot binary code, as the board support package in the GPL source dump is missing.

Rupan commented on target/linux/realtek/image/Makefile:
```diff
@@ -60,11 +60,24 @@ define Device/netgear_gs110tpp-v1
 
+define Device/netgear_gs108t-v3
+  $(Device/Default)
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | uImage lzma
```
Should probably create some sort of shared base device for lzma-enabled U-Boot.

Rupan commented:
I'm actually not entirely sure about the memory size.  U-Boot reports that the system has 128MB of memory, but the markings on the chip indicate that it is 256MB.  The PCB photos on biot's wiki clearly show that the chip is labeled W631GG8MB-12.  The datasheet is here:

https://www.winbond.com/resource-files/W631GG8MB_A03.pdf

The differences between the W631GG8MB variants appear to only involve speed grade and operating temperature.  The datasheet states:

"The W631GG8MB is a 1G bits DDR3 SDRAM, organized as 16,777,216 words x 8 banks x 8 bits."

Assuming that they're using the commonly accepted definition of a word (2 bytes), then ((16777216 * 2 * 8)/1024**2 == 256) is True.  Maybe Netgear ordered a customized SKU and Winbond didn't update the chip markings?

bobafetthotmail commented:
is the dual boot partitions supported in OpenWrt as well? So that when you sysupgrade you flash to the other and then set the uboot to boot from it?

Or OpenWrt uses only one? 

If OpenWrt uses only one how do you switch from booting one image to the other?

bobafetthotmail reviewed (changes requested):
I think you should not delete the generic "default_LZMA" code block.

Can you just include it with  $(Device/Default_LZMA) in Device/Netgear_NGE and then use  $(Device/Netgear_NGE) for the two devices?

Rupan commented:
@bobafetthotmail thank you for the suggestion.  I wasn't sure how many levels of indirection were appropriate in this code base. 
 Since the LZMA definition was only used by the two models I figured I might as well simplify the overall structure.  I've restored the original device definition.

Rupan commented:
@bobafetthotmail regarding the dual partitions, right now I think this is not supported directly by OpenWRT.  I don't yet have a header soldered onto the GS108T serial pads so I don't have ready access to the boot loader.  My understanding is that there is some mechanism in the boot loader, probably via a config variable, to select which partition is booted.

The Netgear GUI allows selection of the boot image, so it should be a matter of dumping the flash before/after a change to see what was modified.

Rupan commented on target/linux/realtek/dts/rtl8380_netgear_gs108t-v3.dts:
```diff
@@ -0,0 +1,202 @@
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "netgear,gs108t-v3", "realtek,rtl838x-soc";
```
I've moved the main definitions into ```rtl8380_netgear_gigabit.dtsi``` following the patterns I see in ```target/linux/ath79/dts```.  I'm not really sure whether I've done it correctly, but the image does build.

bkobl commented:
With regards to the memory size, I have seen it on at least one other Realtek-based switch that u-boot reported only 128MB, while the Ram identified as 256.  There was no adverse effect to making all 256MB available under Linux.

svanheule commented:
Thanks for all the changes so far. Could you also reorder and squash the current commits? You can just force push after doing this.

There's no "best" way to rewrite the changes, but I imagine a first patch to split the device and build definitions, followed by second a patch to add GS108Tv3 support may be enough. Feel free to change it into more or other patches, if you think this is required to better split things into 'logical' changes.

Rupan commented:
I forced-pushed the branch with 3 commits.

bobafetthotmail commented:
knowing what partition is selected is important for sysupgrade, because if you hardcode a partition and then the end user installs OpenWrt in the other, the upgrade will always write to the wrong partition.

You can see how sysupgrade scripts detect what is the active partition from uboot (in a kirkwood-based router) here https://github.com/openwrt/openwrt/tree/master/target/linux/kirkwood/base-files/lib/upgrade

That script needs support for the device's uboot in https://github.com/openwrt/openwrt/tree/master/package/boot/uboot-envtools
that is a tool to read/write uboot configuration from OpenWrt without need to connect serial.

Also the commit that add support of GS108T is missing install instructions

Rupan commented:
I agree that things like sysupgrade need to have proper support added for multi-partition flash like this.  However, the state of this entire platform is currently in-flight and this is probably a decent imcremental improvement.  Will you consider merging as-is?

svanheule commented on target/linux/realtek/image/Makefile:
```diff
@@ -54,17 +61,29 @@ define Device/d-link_dgs-1210-28
+endef
+
+define Device/netgear_gs110tpp-v1
+  $(Device/Netgear_NGE)
```
Could Device/Netgear_NGE be added in the first commit? Then the second commit can contain everything to modify the GS110TPP. The last commit only adds GS108Tv3 support, and nothing else.

Rupan replied:
Sure.  @adschm has left some comments above which I will address in a new commit.  Once there are no further changes requested I will force-push a new commit stack as described.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,126 @@
+};
+
+&spi0 {
+	status = "okay";
```
Please add an empty line after status.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,126 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can this go faster?

Rupan replied:
Not sure; I have seen multiple flash chip vendors for these platforms which probably support different frequencies.  I'll check the data sheet for the one in my switch.

Rupan replied:
From the datasheet: "SPI clock frequencies of W25Q256JV of up to 133MHz are supported".  I did see the response from @svanheule about how this property isn't currently used.  I'll update the speed for now and can remove it if requested.

adschm replied:
Based on @svanheule 's response, I'd be fine with keeping this at 10 MHz ...

Rupan replied:
Honestly I see no reason to revert to 10MHz.  If the node truly is unused then its value is of no consequence.  Both chips known to be in use on this platform -- Macronix MX25L25635F and Winbond W25Q256JV -- support clock frequencies up to 133MHz.  If nothing else than this serves as documentation for future code archeologists.

mans0n replied:
Using untested `spi-max-frequency` is not a good practice, as shown in 9de6bc872af5 and ebc090e420d1.
The spec number is for ideal case; the max frequency on the specific board might be much lower.

Rupan replied:
@mans0n thank you for the review.  Upon closer inspection of the datasheet it turns out that the max frequency should probably be limited to 50MHz, like the other platforms in the referenced commit.  The read data instruction is the limiting factor.

svanheule replied:
I've looked at the binding docs for "jedec,spi-nor", and it turns out spi-max-frequency is (always) required for this node, even if the parent SPI controller does not honour it. 50MHz will _probably_ be fine, although we'll only find out with testing.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,126 @@
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			partition@0 {
```
I'd prefer the partition blocks separated by empty lines (between them and one after size-cells here).

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,126 @@
+			};
+			partition@e0000 {
+				label = "u-boot-env";
+				reg = <0x00e0000 0x010000>;
```
Please use the same number of digits for position and size. The current scheme is made to make it hard to read ...

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gs110tpp-v1.dts:
```diff
@@ -1,226 +1,10 @@
-/dts-v1/;
 
-#include "rtl838x.dtsi"
+/dts-v1/;
```
dts-v1 was wrong in this file beforehand, as it's already in rtl838x.dtsi. Just remove it.

adschm commented:
Actually, the commit message for adding a device should contain specifications and flashing instructions. Please add them.

adschm commented on target/linux/realtek/base-files/etc/board.d/02_network:
```diff
@@ -52,6 +52,8 @@ case $board in
 	ucidef_set_poe 130 "$lan_list"
 	;;
+netgear,gs108t-v3)
+	;;
```
Please remove this useless entry.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gs108t-v3.dts:
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+/dts-v1/;
```
dts-v1 can be dropped.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gs108t-v3.dts:
```diff
@@ -0,0 +1,11 @@
+	compatible = "netgear,gs108t-v3", "realtek,rtl838x-soc";
+	model = "Netgear GS108T v3";
+};
+
```
remove empty line at EOF.

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -19,6 +19,13 @@ define Device/Default
 
+# Use this when U-Boot supports LZMA
+define Device/Default_LZMA
+  $(Device/Default)
```
Device/Default is inherited automatically, and can be dropped.

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -54,17 +61,29 @@ define Device/d-link_dgs-1210-28
-define Device/netgear_gs110tpp-v1
-  $(Device/Default)
+# "NGE" refers to the uImage magic
+define Device/Netgear_NGE
```
I'd personally stick to lower case for the device definition titles.

svanheule commented:
On Sunday, 27 December 2020, Michael Mohr wrote:
> +		spi-max-frequency = <10000000>;
> 
> Not sure; I have seen multiple flash chip vendors for these platforms which probably support different frequencies.  I'll check the data sheet for the one in my switch.

The current driver doesn't even use this property, as the clock speed is fixed. However, it's required for this type of DT node, so that's why it's present here. The SPI driver is currently being rewritten to be able to use spi-memory. As I understand, this won't require the (irrelevant) clock speed specification.

bmork commented:
Wrt flashing instructions:  It is not possible to directly flash an OpenWrt installation with a rootfs from stock firmware.  But it is possible to flash the OpenWrt initramfs image directly from the stock web UI. You can then sysupgrade OpenWrt after booting the initramfs image.

The OpenWrt rootfs parser and sysupgrade will always use the first of the two system partitions, so you should write the initial initramfs to this partition too if you want to keep the stock firmware in the second system partition.  And install the stock firmware in the second of course.

The stock web UI allows direct control of the target partition for both flashing and booting, so switching back and forth between OpenWrt and stock is possible without reinstalling either.  Assuming you can change the bootpartition variable in the second u-boot environment partition (labelled "sysinfo"). See https://patchwork.ozlabs.org/project/openwrt/list/?series=220259 for config.

The "sysinfo" partition should **not** be read-only for this reason.  But the "u-boot-env" could be, actually.  I don't think there is any setting there you'd need or want to change.

bmork commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,126 @@
+			partition@200000{
+				label = "jffs2_log";
+				reg = <0x0200000 0x100000>;
+			};
```
This read-only atttibute usage seems arbitrary? AFAIK, OpenWrt do not access any of these by default.  Why is one of them read-only and not the other?

Rupan replied:
I agree; in fact, I think that most of the partitions should be read-only by default, at least for now.  I've changed all but the main firmware partition to be read-only in the dts (and then only to facilitate sysupgrade).

Rupan commented:
There appears to be no commonality between the existing device trees as to whether the config partitions are read-only.  Given how important the config is to the ability of U-Boot to operate, and the fact that users can still override the read-only setting if they really want to, let's err on the side of caution and set the config partitions to read-only.

ynezz commented:
> Actually, the commit message for adding a device should contain specifications and flashing instructions. Please add them.

@Rupan check b5bd945733a7ed235777c2ceaa0ff27dc779630f for example

Musashii-kun committed:
realtek: add and use netgear_nge for the GS110PP v1

The netgear_nge device will be shared between the GS108T v3 (to be added
in a later commit) and the GS110PP v1.  It also enables LZMA compression
for the ramdisk image.

Signed-off-by: Michael Mohr <akihana@gmail.com>

Musashii-kun committed:
realtek: refactor the Netgear GS110TPP v1 device tree

Move most of the GS110TPP v1 device tree into a dtsi so that it can be
shared with the GS108T v3.  Additionally:

  * Use macros to simplify the ethernet and switch definitions
  * Zero-pad the offsets and sizes in the partition map to 8 digits each

The spi-max-frequency value has been changed from 10MHz to 50MHz based
on an analysis of the relevant datasheets.  The current driver doesn't
use this property, as the clock speed is fixed. However, it's required
for this type of DT node, so that's why it's present here.

The firmware partition has been split in half, since this is how the
stock firmware uses it.  This can be used to easily revert to a stock
firmware if one is written to the second image area.

Signed-off-by: Michael Mohr <akihana@gmail.com>

Musashii-kun committed:
realtek: add support for Netgear GS108T v3

The Netgear GS108T v3 is an 8 port gigabit switch with PoE-PD support
on port 1.  The two prior versions were built using eCos and are not
currently compatible with OpenWRT.

The GS108T v3 is quite similar to both the GS110TPP v1 and GS110TP v3,
all of which use the same firmware image from Netgear.  For this reason,
the device tree is identical aside from the model and compatible values.

All of the above feature a dual firmware layout, referred to as Image0
and Image1 in the Netgear firmware.

Hardware specification
----------------------

 * RTL8380M SoC, 1 MIPS 4KEc core @ 500MHz
 * 128MB DDR3-1600 DRAM (Winbond W631GG8MB-12)
 * 32MB 3v NOR SPI Flash (Macronix MX25L25635F or Winbond W25Q256JVFIQ)
 * RTL8231 GPIO extender to control the LEDs and the reset button
 * 8 x 10/100/1000BASE-T ports, internal PHY (RTL8218B)
 * UART (115200 8N1) via unpopulated standard 0.1" pin header marked J1
 * Power is supplied via a 12V 1A barrel connector or 802.3af

UART pinout
-----------

J1 | [o]ooo
      ^ ||`------ GND
      | |`------- RX         [TX out of the serial adapter]
      | `-------- TX         [RX into the serial adapter]
      `---------- Vcc (3V3)  [the square pin]

The through holes are filled with PB-free solder which melts at 375C.
They can also be drilled using a 0.9mm bit.

Build configuration
-------------------

 * Target System: Realtek MIPS
 * Target Profile: Netgear GS108T v3
 * Target Images -> ramdisk -> Compression: lzma
 * Disable other target images

Boot initramfs image from U-Boot
--------------------------------

 1. Press the Escape key at the `Hit Esc key to stop autoboot` prompt
 2. Init network with `rtk network on` command
 3. Load image with `tftpboot 0x8f000000 openwrt-realtek-generic-netgear_gs108t-v3-initramfs-kernel.bin` command
 4. Boot the image with `bootm` command

The switch defaults to IP 192.168.1.1 and tries to fetch the image via
TFTP from 192.168.1.111.

Updating the installed firmware
-------------------------------

The OpenWRT ramdisk image can be flashed directly from the Netgear UI.
The Image0 slot should be used in order to enable sysupgrade.

As with similar switches, changing the active boot partition can be
accomplished in U-Boot as follows:

 1. Press the Escape key at the `Hit Esc key to stop autoboot` prompt
 2. Run `setsys bootpartition {0|1}` to select the boot partition
 3. Run `savesys` followed by `boota` to proceed with the boot process

Signed-off-by: Michael Mohr <akihana@gmail.com>

Rupan commented:
I rebuilt on top of master and tested netbooting the image.  It works! :)

Rupan commented:
@ynezz I think I've added a fairly thorough writeup in the relevant commit.  Can you take another look?

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -76,6 +76,13 @@ define Device/netgear_gs110tpp-v1
 endef
 TARGET_DEVICES += netgear_gs110tpp-v1
 
+define Device/netgear_gs108t-v3
```
Please apply proper alphabetic sorting.

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -67,15 +67,11 @@ define Device/netgear_nge
   UIMAGE_MAGIC := 0x4e474520
   DEVICE_VENDOR := Netgear
-+endef
+endef
```
This should be done correctly in the original commit, and not fixed later.

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -67,15 +67,11 @@ define Device/netgear_nge
-  DEVICE_VENDOR := NETGEAR
-  DEVICE_MODEL := GS110TP
+  $(Device/netgear_nge)
+  DEVICE_MODEL := GS110TPP
```
This is not about the device tree. Changes should be moved to the first patch as well.

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -19,6 +19,12 @@ define Device/Default
+define Device/default_lzma
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | uImage lzma
+endef
```
I don't think we need this additional definition. I would include the two lines in netgear_nge until we have more users.

adschm commented:
Sorry, this still needs changes.

Rupan commented:
> Sorry, this still needs changes.

Feedback is a gift.  No need to be apologetic :)

Borromini commented:
@Rupan Have you tested the device powered by PoE on OpenWrt? Does that work as well?

Rupan commented:
I have no PoE equipment to test with.

Rupan commented:
It does work when powered by the barrel connector.  I tested netbooting and flashing from the Netgear UI.  Some problem is present with networking after OpenWRT boots up from a netboot, but this doesn't happen when booting from flash.

Borromini commented:
Thanks, bmork confirmed PoE PD works for him.

ynezz commented:
Please rebase.

ynezz commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,137 @@
+
+			partition@300000{
+				label = "firmware";
+				compatible = "netgear,uimage";
```
@Rupan Take a look at this commit ff7709d223e15cfc496abf09acae12826d9216e6 and update this PR accordingly, then ping me, thanks

Rupan replied:
@ynezz thanks for the review.  I've rebased onto ```origin/master``` and updated the device trees.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

ynezz commented:
I'm getting following build error:

```shell
Error: ../dts/rtl8380_netgear_gigabit.dtsi:83.24-25 syntax error
FATAL ERROR: Unable to parse input tree
Makefile:122: recipe for target '/opt/devel/openwrt/openwrt.git/build_dir/target-mips_4kec_musl/linux-realtek_generic/image-rtl8380_netgear_gs108t-v3.dtb' failed
```

adschm commented on target/linux/realtek/image/Makefile:
```diff
@@ -26,6 +26,16 @@ define Device/Default
+  SOC := rtl8380
+  IMAGE_SIZE := 14848k
+  UIMAGE_MAGIC := 0x4e474520
+  DEVICE_VENDOR := Netgear
```
Please use all-caps "NETGEAR" here, as we consistently do in the entire tree.

Rupan replied:
To be clear, ```ZyXEL``` and ```D-Link``` use camel case in the other device definitions in this Makefile.  I'm fine changing it to uppercase if that is the general policy.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,141 @@
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x00000000 0x000e0000>;
```
Would have been easier to read without the useless additional leading zero added here (compared to before copied).

Rupan replied:
I originally had this without the leading zeros, but someone asked me to add them in a previous diff review.  Should I keep them or remove them?

adschm replied:
It's a good idea to have the same number of digits for all partitions, but 7 would be enough here, and adding the eighth number just makes it harder to read again.

Rupan replied:
Sounds good.  I've trimmed the leading zeros (to the highest significant hex digit) for both the flash offset (7)  and size (6).

adschm replied:
But that makes comparing offset and size quite hard actually. Could you please just use 7 digits for all values like it was before you changed the copied source, just like I asked for initially?

Rupan replied:
Sure.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,141 @@
+				read-only;
+			};
+
+			partition@100000{
```
While changing anyway, one could add a space here before the bracket (same below).

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,141 @@
+				label = "firmware2";
+				compatible = "openwrt,uimage", "denx,uimage";
+				openwrt,ih-magic = <0x4e474520>;
+				openwrt,ih-type = <IH_TYPE_KERNEL>;
```
Is having these parameters (and compatible) for two different partitions really a good idea?

ynezz replied:
And what is your suggestion here? Those devices are dual boot.

adschm replied:
And this works OOTB with these parameters now?

ynezz replied:
Yes, it does. You can choose if you would like to boot into OpenWrt or Realtek firmware. By marking it `read-only` for OpenWrt you prevent any accidental modifications of vendor/backup firmware.

adschm replied:
Well, that's nice, wasn't aware of that.

bmork replied:
I don't think that's a good idea. There is no dual image support for this target in OpenWrt, so rootfs splitting can only work for one of the partitions.  We should make it explicit which one that is, and so far it has been the first one.

This dts might result in OpenWrt booting from one partition and mounting the rootfs from the other. 

FWIW, I did consider support for using both partitions.  But AFAICS, there is no reliable way to figure out which one we are booting from. The bootpartition variable can be modified in u-boot memeory without writing it back to the sysinfo partition. This might be rare enough to be ignored, if dual partition support was important.  But I don't think it is.  It's just complicating stuff unnecessarily.  So I opted for OpenWrt in the first, and optionally stock in the second image.  Makes everything simple.

This is not unique for this target.  The Unifi AC family has similar separation.

What's unique about these realtek devices is that the stock web application allows directly selecting which partition to flash and boot.  This means that there is no problem reserving one of the partitions for OpenWrt, even if you boot to stock now and then.  You can even upgrade stock without issues.

Rupan replied:
I don't think it is a good idea to make the primary firmware partition the size of the spare flash, as is the case in the current DTS.  Someone might unknowingly overwrite the secondary firmware area.  I see a few possibilities:

1) The partition definition @ 1180000 could just be removed, leaving the partition definition @ 300000 as-is.
2) Keep the second firmware partition as-is, with its read-only attribute.  This could serve as a signal for which partition is booting (the only one that is read-write).
3) Remove the compatible and ih-magic properties and name it something like "reserved".

Any preferences?

bmork replied:
The partition layout should match what stock firmware has.  I did not know there was a current DTS with something else.

I don't think the rootfs splitting code cares about read-only status, so it might still be confused if both partitions have the openwrt,uimage compatible.  IMHO, it should only be used for partitions we can boot OpenWrt from.

So I prefer 3.  As for the names, I would personally have preferred that we used something closer to stock u-boot and firmware for most of the partitions (except "firmware" obviously, since that is special to sysupgrade).  This is the table from u-boot on my GS108Tv3:
```
RTL838x# flshow
=============== FLASH Partition Layout ===============
Index  Name       Size       Address
------------------------------------------------------
 0     LOADER     0xe0000    0xb4000000-0xb40dffff
 1     BDINFO     0x10000    0xb40e0000-0xb40effff
 2     SYSINFO    0x10000    0xb40f0000-0xb40fffff
 3     JFFS2_CFG  0x100000   0xb4100000-0xb41fffff
 4     JFFS2_LOG  0x100000   0xb4200000-0xb42fffff
 5     RUNTIME1   0xe80000   0xb4300000-0xb517ffff
 6     RUNTIME2   0xe80000   0xb5180000-0xb5ffffff
======================================================
```

The stock firmware also use almost those names when relevant.  Example boot log:

```
Probe: SPI CS1 Flash Type MX25L12845E
Creating 7 MTD partitions on "Total SPI FLASH":
0x00000000-0x00040000 : "LOADER"
0x00040000-0x00050000 : "BDINFO"
0x00050000-0x00060000 : "SYSINFO"
0x00060000-0x00160000 : "JFFS2 CFG"
0x00160000-0x00260000 : "JFFS2 LOG"
0x00260000-0x00930000 : "RUNTIME"
0x00930000-0x01000000 : "RUNTIME2"
```
The differences "RUNTIME" vs "RUNTIME1" and space instead of underscore is still interesting....  But the second image partition is consistently named  "RUNTIME2".  I believe we should use the same name, possibly in lower case.  This will be least confusing to users switching between OpenWrt, u-boot and stock.  

But

Rupan replied:
That all sounds reasonable to me.  I've renamed the partitions in the DTSI to match what the stock firmware uses (aside from the primary runtime partition) and removed the compatible/ih-magic properties from runtime2.

adschm commented on target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi:
```diff
@@ -0,0 +1,141 @@
+				label = "firmware";
+				compatible = "openwrt,uimage", "denx,uimage";
+				openwrt,ih-magic = <0x4e474520>;
+				openwrt,ih-type = <IH_TYPE_KERNEL>;
```
Looks like ih-type is what causes the syntax error. And that's also the change you introduce which wasn't there before.

ynezz replied:
It's due to missing #include <dt-bindings/mtd/partitions/uimage.h>

svanheule replied:
I didn't test this, but I think this breaks due to a missing include of "dt-bindings/mtd/partitions/uimage.h", introduced in dd5ac0546ca3f80de10db309e57090cf809b1b72

bmork replied:
yes, all these properties are meant only for overriding defaults.  No need to repeat default values in the dts

Rupan commented:
Whoops, sorry about the build error.  I was a bit tired last night and started the build at the same time I force-pushed the branch.

Rupan commented:
Sorry again about the broken dtsi.  I successfully built an image with this branch just now.

Rupan commented:
I re-confirmed that the build still works :) @ynezz - look good now?

Rupan commented:
The code -- in its current form -- does build.  Given the number of change requests I will not test flashing on real hardware again until the code is tagged as ready for merge.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

Rupan commented:
Thank you!

Neustradamus commented:
Merged commits:
- https://github.com/openwrt/openwrt/commit/9877393fd135f247f58bd4dc6f1532f77f1c40c8
- https://github.com/openwrt/openwrt/commit/9fe90e4e5aefb1f23d2f23b1e6bcd20f3b548c09
- https://github.com/openwrt/openwrt/commit/8cbd6f51174995c639dfb14faed46b4ba33542b8
